joomla/joomla-cms
# [com_fields] fix with deleting data on item update, forms checkbox fix

### Summary of Changes

Main concept is:

1. `checkbox` and `checkboxes` fields have a special property `$ensure_value` to indicate that these fields require any value.

`JForm::filter()` method now accepts `$ensure_values` param enabled by default, hence the filtered data always have values for such fields (hence saving empty checkbox will have data)

2. Custom fields plugin only updates the value which really exists in `$data['com_fields']`

Result: no custom fields data is lost on items save and checkboxes fine in forms as well.

No BC breaks.

Related to #18188, #17837, #17801.

### Testing Instructions
Save users/articles with checkboxes custom fields.

Ensure that checked, unchecked data is saved.

Save user profiles, reset password, block/unblock/activate users, ensure that no custom fields values are lost.

Perform direct `JUser::save()` related operations, ensure that no custom fields values are lost:

```
$user = JFactory::getUser($USER_ID);
$user->save();

JUserHelper::addUserToGroup();
JUserHelper::removeUserFromGroup();
````
